modules/twinkleblock.js

```diff
@@ -0,0 +1,879 @@
+
+  if ($form.find('[name=actiontype][value=block]').is(':checked')) {
+    field_preset = new Morebits.quickForm.element({ type: 'field', label: 'Preset', name: 'field_preset' });
+    field_preset.append({
+        type: 'select',
+        name: 'category',
+        label: 'Choose a preset:',
+        event: Twinkle.block.callback.change_preset,
+        list: [{
+          label: '',
+          value: '',
+          type: 'option'
+        }].concat(Twinkle.block.callback.filtered_block_groups())
+      });
+
+    field_block_options = new Morebits.quickForm.element({ type: 'field', label: 'Block options', name: 'field_block_options' });
+    field_block_options.append({ type: 'div', name: 'currentblock', label: ' ' });
+    field_block_options.append({
+        type: 'select',
+        name: 'expiry_preset',
+        label: 'Expiry:',
+        event: Twinkle.block.callback.change_expiry,
+        list: [
+          { label: 'custom', value: 'custom', selected: true },
+          { label: 'indefinite', value: 'infinity' },
+          { label: '3 hours', value: '3 hours' },
+          { label: '12 hours', value: '12 hours' },
+          { label: '24 hours', value: '24 hours' },
+          { label: '31 hours', value: '31 hours' },
+          { label: '36 hours', value: '36 hours' },
+          { label: '48 hours', value: '48 hours' },
+          { label: '60 hours', value: '60 hours' },
+          { label: '72 hours', value: '72 hours' },
+          { label: '1 week', value: '1 week' },
+          { label: '2 weeks', value: '2 weeks' },
+          { label: '1 month', value: '1 month' },
+          { label: '3 months', value: '3 months' },
+          { label: '6 months', value: '6 months' },
+          { label: '1 year', value: '1 year' },
+          { label: '2 years', value: '2 years' },
+          { label: '3 years', value: '3 years' }
+        ]
+      });
+    field_block_options.append({
+        type: 'input',
+        name: 'expiry',
+        label: 'Custom expiry',
+        tooltip: 'You can use relative times, like "1 minute" or "19 days", or absolute timestamps, "yyyymmddhhmm" (e.g. "200602011405" is Feb 1, 2006, at 14:05 UTC).'
+      });
+    var blockoptions = [
+        {
+          label: 'Block account creation',
+          name: 'nocreate',
+          value: '1'
+        },
+        {
+          label: 'Block user from sending email',
+          name: 'noemail',
+          value: '1'
+        },
+        {
+          label: 'Prevent this user from editing their own talk page while blocked',
+          name: 'disabletalk',
+          value: '1'
+        }
+      ];
+
+    if (Twinkle.block.isRegistered) {
+      blockoptions.push({
+          label: 'Autoblock any IP addresses used',
+          name: 'autoblock',
+          value: '1'
+        });
+    } else {
+      blockoptions.push({
+          label: 'Prevent logged-in users from editing from this IP address',
+          name: 'hardblock',
```

Comment:
Feel free to use WP jargon in the label if you think it would be clearer. I don't deal much in blocking, but if everyone thinks of this as a "hard block", then the label in Twinkle could be "Hard block (prevent logged-in users from editing from this IP address)". Up to you.